src/clp_ffi_js/ir/ClpIrV1Decoder.cpp

```diff
@@ -0,0 +1,233 @@
+        throw ClpJsException(
+                clp::ErrorCode::ErrorCode_MetadataCorrupted,
+                __FILENAME__,
+                __LINE__,
+                "Failed to decode encoding type."
+        );
+    }
+
+    if (false == is_four_bytes_encoding) {
+        throw ClpJsException(
+                clp::ErrorCode::ErrorCode_Unsupported,
+                __FILENAME__,
+                __LINE__,
+                "Is not four byte encoding."
+        );
+    }
+    auto result{clp::ir::LogEventDeserializer<clp::ir::four_byte_encoded_variable_t>::create(
+            *zstd_decompressor
+    )};
+    if (result.has_error()) {
+        auto const error_code{result.error()};
+        SPDLOG_CRITICAL(
+                "Failed to decompress: {}:{}",
+                error_code.category().name(),
+                error_code.message()
+        );
+        throw ClpJsException(
+                clp::ErrorCode::ErrorCode_MetadataCorrupted,
+                __FILENAME__,
+                __LINE__,
+                "Failed to decompress"
+        );
+    }
+
+    return ClpIrV1Decoder(std::move(data_buffer), zstd_decompressor, std::move(result.value()));
+}
+
+auto ClpIrV1Decoder::get_estimated_num_events() const -> size_t {
+    return m_log_events.size();
+}
+
+auto ClpIrV1Decoder::build_idx(size_t begin_idx, size_t end_idx) -> emscripten::val {
+    constexpr size_t cFullRangeEndIdx{0};
+    if (0 != begin_idx && cFullRangeEndIdx != end_idx) {
+        throw ClpJsException(
+                clp::ErrorCode::ErrorCode_Unsupported,
+                __FILENAME__,
+                __LINE__,
+                "Partial range index building is not yet supported."
+        );
+    }
+    if (false == m_full_range_built) {
+        m_full_range_built = true;
+        constexpr size_t cDefaultNumLogEvents{500'000};
+        m_log_events.reserve(cDefaultNumLogEvents);
+        while (true) {
+            auto const result{m_deserializer.deserialize_log_event()};
+            if (false == result.has_error()) {
+                m_log_events.emplace_back(result.value());
+                continue;
+            }
+            auto const error{result.error()};
+            if (std::errc::no_message_available == error) {
+                break;
+            }
+            if (std::errc::result_out_of_range == error) {
+                SPDLOG_ERROR("File contains an incomplete IR stream");
+                break;
+            }
+            throw ClpJsException(
+                    clp::ErrorCode::ErrorCode_Corrupt,
+                    __FILENAME__,
+                    __LINE__,
+                    "Failed to decompress: "s + error.category().name() + ":" + error.message()
+            );
+        }
+        m_data_buffer.reset(nullptr);
+    }
+
+    auto results{emscripten::val::object()};
+    results.set("numValidEvents", m_log_events.size());
+    results.set("numInvalidEvents", 0);
+    return results;
```

Comment:
Since we're not using `numInvalidEvents`, how about let's keep it simple and just return the number of events buffered so far. As mentioned in the other comment, we can wrap this library in the new log viewer to provide the necessary decoder API.